index.js

```diff
@@ -1,47 +1,246 @@
+    if (info.channel) {
+      lockedAdd(info.channel)
+    }
+
+    // add any requested dats
+    stream.on('feed', lockedAdd)
+
+    function lockedAdd (dkey) {
+      self._lock(release => {
+        _add(dkey)
+          .then(() => release())
+          .catch(err => release(err))
+      })
+    }
+
+    async function _add (dkey) {
+      const dkeyStr = datEncoding.toStr(dkey)
+
+      // lookup the archive
+      try {
+        var feed = self._replicatingFeeds.get(dkeyStr)
+        if (!feed) return // TODO: error ?
+      } catch (err) {
+        if (!stream.destroyed) stream.destroy(err)
+      }
+
+      self._replicatingFeeds.set(dkeyStr, feed)
+
+      if (!feed || !feed.isSwarming) {
+        return
+      }
+
+      if (!feed.replicationStreams) {
+        feed.replicationStreams = []
+      }
+      if (feed.replicationStreams.indexOf(stream) !== -1) {
+        return // already replicating
+      }
+
+      // create the replication stream
+      feed.replicate({ stream, live: true })
+      if (stream.destroyed) return // in case the stream was destroyed during setup
+
+      // track the stream
+      var keyStr = datEncoding.toStr(feed.key)
+      streamKeys.push(keyStr)
+      feed.replicationStreams.push(stream)
+
+      function onend () {
+        feed.replicationStreams = feed.replicationStreams.filter(s => (s !== stream))
+
+        // If the Replicator is the only object with a reference to this core, close it after replication's finished.
+        if (!feed.replicationStreams.length) {
+          self._replicatingFeeds.delete(dkeyStr)
+          feed.close()
+        }
+      }
+      stream.once('error', onend)
+      stream.once('end', onend)
+      stream.once('close', onend)
+    }
+
+    // debugging
+    stream.on('error', err => {
+      debug({
+        event: 'connection-error',
+        peer: `${info.host}:${info.port}`,
+        connectionType: info.type,
+        message: err.toString()
+      })
+    })
+
+    return stream
+  }
+
+  add (feed) {
+    if (!feed.key) return feed.ready(() => { this.add(feed) })
+    const key = datEncoding.toStr(feed.key)
+    const discoveryKey = datEncoding.toStr(feed.discoveryKey)
+    this._replicatingFeeds.set(discoveryKey, feed)
+
+    this.rejoin(feed.discoveryKey)
+    this.emit('join', { key, discoveryKey })
+    feed.isSwarming = true
+  }
+
+  rejoin (discoveryKey) {
+    this._swarm.join(discoveryKey)
+  }
+
+  listen (port) {
+    port = port || this.port
+    this._swarm.listen(port)
+    return new Promise(resolve => {
+      this._swarm.once('listening', resolve)
+    })
+  }
+
+  leave (discoveryKey) {
+    const feed = this._replicatingFeeds.get(discoveryKey)
+    if (!feed) return
+    if (feed.replicationStreams) {
+      feed.replicationStreams.forEach(stream => stream.destroy()) // stop all active replications
+      feed.replicationStreams.length = 0
+    }
+    this._swarm.leave(discoveryKey)
+  }
+
+  async close () {
```

Comment:
The `async` keyword is counter-productive, both here an inside the the promise callback.